### PR TITLE
refactor(broker): generate event keys in stream processor

### LINKED ISSUE #1053: Stream processing: generate keys in stream processors
Currently, we generate record keys as the position on the log when the initiating command/event is published. This means that at the time of publishing we do not know the key. However, this is desired in order to store things in the index related to that new record (e.g. activity instance keys during workflow processing).

AT:

* Stream processors use a key generator when writing new events that is part of the snapshotted state
* Uniqueness of keys on a partition is ensured (i.e. we currently have three entity-managing stream processors that run independently; they should use different key spaces; we will merge those processors into one stream processor in the future, and then we can have a single key space)
* Commands that create a new entity have a key == -1 (because we cannot assign keys from the outside anymore); responses to those commands must contain the key

This is a minor protocol change, because the `key` field is no longer populated, when a command is received via a topic subscription.

### PR DESCRIPTION
**Pull request by ThorbenLindhauer**
- allows the stream processor to use the key immediately after
  publication, e.g. as a key in an index map (as compared to #positionAsKey,
  where we cannot know the key in the publishing stream processor)
- writes snapshots for the key counter
- ensures that keys for wf instances, jobs, incidents on the same
  partition are unique by giving them a separate key space; as soon as
  we merge their stream processors, we can have a single key space and
  key generator
- protocol changes:
  - commands that create a new entity (e.g. workflow instance) get no
    key assigned, i.e. the command's key is -1
  - command rejection in these cases has no key either
  - the topic record with intent CREATE_COMPLETE is now a command
    instead of an event; this makes more sense as it is published from
    the outside and can be followed by rejection

closes #1053

### CODE REVIEW

**Comment by menski** [comment]
I rebased the branch on develop to fix the merge conflict

File: broker-core/src/main/java/io/zeebe/broker/subscription/message/MessageService.java [hunk 1]
```diff
@@ -63,6 +64,7 @@ private TypedStreamProcessor createStreamProcessors(TypedStreamEnvironment env)
     final MessageDataStore messageDataStore = new MessageDataStore();
 
     return env.newStreamProcessor()
+        .keyGenerator(new KeyGenerator(0, 1))
```
Comment (saig0) [inline broker-core/src/main/java/io/zeebe/broker/subscription/message/MessageService.java]: In order to ensure unique keys, I guess that the message stream processors also need keys with offset from `KeyGenerator`.

File: broker-core/src/main/java/io/zeebe/broker/incident/processor/IncidentStreamProcessor.java [hunk 2]
```diff
@@ -54,12 +55,16 @@
...
+
     TypedEventStreamProcessorBuilder builder =
         env.newStreamProcessor()
+            .keyGenerator(KeyGenerator.createIncidentKeyGenerator())
             .withStateResource(activityInstanceMap)
             .withStateResource(failedJobMap)
             .withStateResource(incidentMap.getMap())
-            .withStateResource(resolvingEvents);
+            .withStateResource(resolvingEvents)
+            .withStateResource(keyGenerator);
```
Comment (saig0) [inline broker-core/src/main/java/io/zeebe/broker/incident/processor/IncidentStreamProcessor.java]: You don't need to register the `KeyGenerator` as a state resource. This is already done by setting it.

**Review by saig0 — CHANGES_REQUESTED** [review]

File: broker-core/src/main/java/io/zeebe/broker/subscription/message/MessageService.java (hunk 1 above)
Comment (ThorbenLindhauer) [reply broker-core/src/main/java/io/zeebe/broker/subscription/message/MessageService.java]: Not sure we need this at this point. These records are not exposed to users via topic subscriptions and internally we also do not require uniqueness currently.

**Comment by ThorbenLindhauer** [comment]
I addressed the other change request.

**Review by saig0 — APPROVED** [review]

**Pull request merged**